### DISCUSSION TITLE
How to apply a theme conditionally (e.g. when available in context)

### DISCUSSION THREAD

**Discussion by EliotRagueneau** [molstar/molstar#1106] [Q&A]
I want to apply the `PLDDTConfidenceColorTheme` when my structure is coming from AlphaFold, and therefore my structure has pLDDT metrics, but otherwise use a uniform #888888 color theme.

I am in an headless plugin context, so to get access to the color-theme, i first do
```ts
    const defaultSpec = DefaultPluginSpec();
    const plugin = new HeadlessPluginContext(externalModules, {
            ...defaultSpec,
            behaviors: [
                PluginSpec.Behavior(MAQualityAssessment, {autoAttach: true}), 
                ...defaultSpec.behaviors
            ]
        }, {width: 800, height: 800},
        {...}
    );
```

So far, the closest I've been to making it work has been to update the default color theme of the representation I am using like the following, before creating the representation:
```ts
    CartoonRepresentationProvider.defaultColorTheme.name = 'uniform';
    CartoonRepresentationProvider.defaultColorTheme.props = {value: Color(0x888888)};
```

Then I actually use the rendering:
```ts
const update = plugin.build();
await update.toRoot()
        .apply(Download, {url})
        .apply(TrajectoryFromPDB)
        .apply(ModelFromTrajectory)
        .apply(StructureFromModel, {type: {name: 'assembly', params: {}}})
        .apply(StructureComponent, {type: {name: 'static', params: 'polymer'}});
        .apply(StructureRepresentation3D, {
            type: {name: 'cartoon', params: {alpha: 1, ignoreLight: true}},
            colorTheme: {name: 'plddt-confidence', params: {}},
    }).commit();
```

However, there are 2 big downsides of this:
1. The logic is rather restraint as we are relying on the failure of `plddt-confidence` to then apply the default. It therefore works in this case, but wouldn't if the color-theme I wanted to apply conditionally was working all the time. Also, I find it unclean to modify the default ColorTheme of Cartoon like this
2. It actually doesn't work since the defaultColorTheme.props are not taken into account

My question is therefore the following: what is the best way to apply a theme conditionally ? Maybe using a Transformer ?

**Comment by dsehnal** [comment]
Hi, something like this should work and no need to change the defaults.

```ts
const structure = await plugin.build().toRoot()
        .apply(Download, {url})
        .apply(TrajectoryFromPDB)
        .apply(ModelFromTrajectory)
        .apply(StructureFromModel, {type: {name: 'assembly', params: {}}})
        .commit();

if (!structure.data) throw new Error(`Failed to create structure: {structure.cell?.errorText}`);

const hasScores = structure.data.models.some(m => QualityAssessment.isApplicable(m, 'pLDDT'); 
await plugin.build()
  .to(structure)
  .apply(StructureComponent, ...);
  .apply({ ..., colorTheme: hasScores ? {} : {} })
  .commit();
```

**Reply by EliotRagueneau** [reply]
Oh I didn't noticed that commit was returning something, that is extremely useful!

Thanks a lot!

**Reply by dsehnal** [reply]
You can get a selector for any part of the tree during the build phase, e.g.

```ts
const update = plugin.build();
const a = update.toRoot().apply(...)
const b = a.apply(...)

return [a.selector, b.selector];
```
